runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -409,6 +428,46 @@ function LanguageTree:_add_injections()
   return query_time
 end
 
+--- Run an asynchronous parse, calling {on_parse} when complete.
+---
+--- @param range boolean|Range|nil
+--- @param on_parse fun(trees: TSTrees)
+function LanguageTree:_async_parse(range, on_parse)
+  coroutine.resume(
+    --- @param co_range boolean|Range|nil
+    --- @param co_cb fun(trees: TSTrees)
+    coroutine.create(function(co_range, co_cb)
+      local co = coroutine.running()
+      local ct = self:_buf().changedtick
+      local unfinished = true
+      --- @type TSTrees
+      local trees
+
+      while unfinished do
+        -- If buffer was changed in the middle of parsing, reset parse state
+        if self:_buf().changedtick ~= ct then
+          self._parser:reset()
+          self:invalidate()
+          coroutine.yield()
+        end
+
+        trees, unfinished = self:_parse(co_range, default_parse_timeout_ms)
+
+        if unfinished then
+          vim.schedule(function()
+            coroutine.resume(co)
+          end)
+          coroutine.yield()
+        else
+          co_cb(trees)
+        end
+      end
+    end),
+    range,
```

Comment:
Since this coroutine is a closure, wouldn't it be better to not pass these in and use upvalues?